Support IOS setup in AppDelegate.m

jonastelzio opened the pull request:
On IOS, CallKit doesn't get it's delegate registered before actively calling .setup from react native, meaning that we'll never get any didLoadWithEvents fired for events such as answer and end, since CallKit simply isn't delegated to callkeep prior answering the call.

This PR exposes setup so it can be called from AppDelegate.m, allowing us to catch these events as soon as they happen.

Example of how to perform setup during AppDelegate initialization:
```
@implementation AppDelegate
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{ 
  self.bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];

  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:self.bridge
                                                   moduleName:@"App"
                                            initialProperties:nil];

  // ======== OTHER CODE REDACTED ==========

  // RNCallKeep setup should (probably?) be called before registering for PuskKit notifications
  [RNCallKeep setup:@{
    @"appName": @"Awesome App",
    @"maximumCallGroups": @3,
    @"maximumCallsPerCallGroup": @1,
    @"supportsVideo": @NO,
  }];

  // Register for pushkit. As soon as the delegates are registered, Pushkit events will fire.
  // It probably(?) doesn't happen syncronously, but RNCallKeep should probably be registered first
  [RNVoipPushNotificationManager voipRegistration];

  // ======== OTHER CODE REDACTED ==========
  
  return YES;
}
```

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -54,8 +55,10 @@ - (instancetype)init
     if (self = [super init]) {
-        _isStartCallActionEventListenerAdded = NO;
-        _delayedEvents = [NSMutableArray array];
+        if (_delayedEvents == nil) {
```
init gets called several times, each time nuking this collection, resulting in us not getting any events at all.

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -118,10 +121,10 @@ - (void)sendEventWithNameWrapper:(NSString *)name body:(id)body {
+        NSDictionary *dictionary = [NSDictionary dictionaryWithObjectsAndKeys:
+                              name, @"name",
+                              body, @"data",
+                              nil];
```
Certain events have `data` as null - this approach gets rid of an exception when that's the case.

Romick2005 commented:
I did the same in my PR https://github.com/react-native-webrtc/react-native-callkeep/pull/382#issuecomment-806785809

jonastelzio commented:
> I did the same in my PR [#382 (comment)](https://github.com/react-native-webrtc/react-native-callkeep/pull/382#issuecomment-806785809)

That pull request is really difficult to review because you changed indentation so many places in the file ;)

Romick2005 commented:
I also don't care what you are thinking. I just prefer have better code in the repo that I need to use and that's it.

danjenkins commented:
That's a great way to talk to one of the people that would approve your PR...

jonastelzio commented on ios/RNCallKeep/RNCallKeep.h:
```diff
@@ -45,4 +45,6 @@ continueUserActivity:(NSUserActivity *)userActivity
 + (BOOL)isCallActive:(NSString *)uuidString;
 
++ (void)setup:(NSDictionary *)options;
+
```
Expose a static method to call setup from `AppDelegate,m`

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -41,6 +41,7 @@ @implementation RNCallKeep
     BOOL _isStartCallActionEventListenerAdded;
     bool _hasListeners;
     NSMutableArray *_delayedEvents;
+    bool _isSetup;
```
Track if we've already called setup.

I assume it's not an existing feature to change the setup of this library after it's initally been configured (?)

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -55,7 +56,7 @@ - (instancetype)init
     if (self = [super init]) {
         _isStartCallActionEventListenerAdded = NO;
-        _delayedEvents = [NSMutableArray array];
+        if (_delayedEvents == nil) _delayedEvents = [NSMutableArray array];
```
`init` is called repeatedly, deleting any events queued up here. This avoids that.

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -118,10 +119,11 @@ - (void)sendEventWithNameWrapper:(NSString *)name body:(id)body {
+            name, @"name",
+            body, @"data",
+            nil
+        ];
```
`body` can be null for certain events, such as `RNCallKeepDidActivateAudioSession` which will fire before the bridge is up if answering a call really quickly!

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -133,8 +135,20 @@ + (void)initCallKitProvider {
++ (void)setup:(NSDictionary *)options {
+    RNCallKeep *callKeep = [RNCallKeep allocWithZone: nil];
+    [callKeep setup:options];
+}
```
Simply allocate the singleton and call setup in there.

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -133,8 +135,20 @@ + (void)initCallKitProvider {
+#endif
+        return;
+    }
+    
```
If we've already called setup, we ignore subsequent calls to this method.

I'm not sure if this is something we'd want. Is it expected to be able to change the setup during the life cycle?

danjenkins replied:
I wouldnt say so

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -183,7 +199,7 @@ + (void)initCallKitProvider {
                          supportsDTMF:(BOOL)supportsDTMF
                      supportsGrouping:(BOOL)supportsGrouping
-                   supportsUngrouping:(BOOL)supportsUngrouping)            
+                   supportsUngrouping:(BOOL)supportsUngrouping)
```
Whatever was sitting at the end of this line, xcode insists on deleting it when I save the file 🤷‍♂️

danjenkins replied:
whitespace is the devil

jonastelzio commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-native-callkeep",
-  "version": "4.1.0",
+  "version": "4.1.1-iosfix",
```
Version bump so I can use this myself while we're reviewing. I assume we'd just change this when merging(?)

jonastelzio commented:
Upon further investigation, it might be sufficient to just call [RNCallKeep initCallKitProvider] instead of setup, considering it's using `NSUserDefaults` to obtain these settings.

I'll make some testing and amend the PR if that's the case.

danjenkins commented:
@jonastelzio can you make changes to the README too? Other than that I've confirmed these changes work in a project so I'd be happy to approve once theres docs on using setup in app delegate

jonastelzio commented:
> @jonastelzio can you make changes to the README too? Other than that I've confirmed these changes work in a project so I'd be happy to approve once theres docs on using setup in app delegate

Can do!

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -133,8 +136,22 @@ + (void)initCallKitProvider {
+    RNCallKeep *callKeep = [RNCallKeep allocWithZone: nil];
+    [callKeep setup:options];
+    isSetupNatively = YES;
+}
```
Simply allocate the singleton and call setup in there, setting isSetupNatively makes the library ignore further calls to setup from react native.

jonastelzio commented on ios/RNCallKeep/RNCallKeep.m:
```diff
@@ -133,8 +136,22 @@ + (void)initCallKitProvider {
+        NSLog(@"[RNCallKeep][setup] already setup");
+        RCTLog(@"[RNCallKeep][setup] already setup in native code");
+#endif
+        return;
```
If we've already initialized from `AppDelegate.m`, we ignore further calls to setup from React Native.

jonastelzio commented:
@danjenkins I've added documentation and changed the behavior so we ONLY prevent repeated calls to .setup from react native IF the module is initially setup from native code. This way this push will never change any existing functionality.

Jerome91410 commented:
already discussed in
https://github.com/react-native-webrtc/react-native-callkeep/pull/365

And no feedback since 24 days...

gobzila commented:
Any updates on this? It would be great if this PR will be merged.

manuquentin commented:
Thanks @jonastelzio !
